contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.2.0--3.3.0.sql

```diff
@@ -424,6 +424,375 @@ END
+	-- But when we copy the #tmp table into another table, we do get all the attributes and metadata
+	DECLARE @tmptab VARCHAR(63) = 'sp_babelfish_autoformat' + REPLACE(NEWID(), '-', '')
+	DECLARE @tmptab2 VARCHAR(63) = 'sp_babelfish_autoformat' + REPLACE(NEWID(), '-', '')
+	DECLARE @cmd VARCHAR(1000) = 'SELECT * INTO ' + @tmptab + ' FROM ' + @tab
+	EXECUTE(@cmd)
+
+	-- Get the columns
+	SELECT 
+	   c.name AS colname, c.colid AS colid, t.name AS basetype, 0 AS maxlen
+	INTO #sp_bbf_autoformat
+	FROM sys.syscolumns c left join sys.systypes t 
+	ON c.xusertype = t.xusertype		
+	WHERE c.id = object_id(@tmptab)
+	ORDER BY c.colid
+
+	-- Get max length for each column based on the data
+	DECLARE @colname VARCHAR(63), @basetype VARCHAR(63), @maxlen int
+	DECLARE c CURSOR FOR SELECT colname, basetype, maxlen FROM #sp_bbf_autoformat ORDER BY colid
+	OPEN c
+	WHILE 1=1
+	BEGIN
+		FETCH c INTO @colname, @basetype, @maxlen
+		IF @@fetch_status <> 0 BREAK
+		SET @cmd = 'DECLARE @i INT SELECT @i=ISNULL(MAX(LEN(CAST([' + @colname + '] AS VARCHAR(500)))),4) FROM ' + @tmptab + ' UPDATE #sp_bbf_autoformat SET maxlen = @i WHERE colname = ''' + @colname + ''''
+		EXECUTE(@cmd)
+	END
+	CLOSE c
+	DEALLOCATE c
+
+	-- Generate the final SELECT
+	DECLARE @selectlist VARCHAR(8000) = ''
+	DECLARE @collist VARCHAR(8000) = ''
+	DECLARE @fmtstart VARCHAR(30) = ''
+	DECLARE @fmtend VARCHAR(30) = ''
+	OPEN c
+	WHILE 1=1
+	BEGIN
+		FETCH c INTO @colname, @basetype, @maxlen
+		IF @@fetch_status <> 0 BREAK
+		IF LEN(@colname) > @maxlen SET @maxlen = LEN(@colname)
+		IF @maxlen <= 0 SET @maxlen = 1
+		
+		IF (CHARINDEX(',' + UPPER(@colname) + ',', @hiddencols) > 0) OR (CHARINDEX(',[' + UPPER(@colname) + '],', @hiddencols) > 0) 
+		BEGIN
+			SET @selectlist += ' [' + @colname + '],'			
+		END
+		ELSE 
+		BEGIN
+			SET @fmtstart = ''
+			SET @fmtend = ''
+			IF @basetype IN ('tinyint', 'smallint', 'int', 'bigint', 'decimal', 'numeric', 'real', 'float') 
+			BEGIN
+				SET @fmtstart = 'CAST(right(space('+CAST(@maxlen AS VARCHAR)+')+'
+				SET @fmtend = ','+CAST(@maxlen AS VARCHAR)+') AS VARCHAR(' + CAST(@maxlen AS VARCHAR) + '))'
+			END
+
+			SET @selectlist += ' '+@fmtstart+'CAST([' + @colname + '] AS VARCHAR(' + CAST(@maxlen AS VARCHAR) + '))'+@fmtend+' AS [' + @colname + '],'
+			SET @collist += '['+@colname + '],'
+		END
+	END
+	CLOSE c
+	DEALLOCATE c
+
+	-- Remove redundant commas
+	SET @collist = SUBSTRING(@collist, 1, LEN(@collist)-1)
+	SET @selectlist = SUBSTRING(@selectlist, 1, LEN(@selectlist)-1)	
+	SET @selectlist = 'SELECT ' + @selectlist + ' INTO ' + @tmptab2 + ' FROM ' + @tmptab + ' ' + @orderby
+	EXECUTE(@selectlist)
+	EXECUTE('SELECT ' + @collist + ' FROM ' + @tmptab2)
+	
+	-- PRINT rowcount
+	SET @rc = @@rowcount
+	IF @printrc = 1
+	BEGIN
+		PRINT '   '
+		SET @cmd = '(' + CAST(@rc AS VARCHAR) + ' rows affected)'
+		PRINT @cmd
+	END
+	
+	-- Cleanup
+	EXECUTE('DROP TABLE ' + @tmptab)
+	EXECUTE('DROP TABLE ' + @tmptab2)
```

Comment:
Should we be executing the entire logic inside a try/catch? Incase of unwanted/unknown errors before dropping, the tables could be left stale

Reply:
I guess we could put SET XACT_ABORT OFF at the top of the proc, so that execution still reaches the end of the procedure (where the DROP stmts are) in case of errors.

Reply:
`Only an error that would abort the proc execution would miss those DROPs, but I think that is also true for TRY-CATCH`: Any error aborting in the body of the try will take the control flow to the catch block.

`One thing we could do is add a timestamp to the table name and drop any such tables older than, say, 24 hours. But I wonder if all that is worth it.` : 
Drawbacks: 1. Might be a potential to Bloating the storage. 2. We would need to run some cleanup